connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/inbound/DefaultProcessInstanceContext.java

```diff
@@ -0,0 +1,105 @@
+import io.camunda.connector.api.inbound.ProcessInstanceContext;
+import io.camunda.connector.api.validation.ValidationProvider;
+import io.camunda.connector.api.validation.ValidationUtil;
+import io.camunda.connector.feel.FeelEngineWrapper;
+import io.camunda.connector.feel.FeelEngineWrapperException;
+import java.util.HashMap;
+import java.util.Map;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class DefaultProcessInstanceContext implements ProcessInstanceContext {
+  private static final Logger LOGGER = LoggerFactory.getLogger(DefaultProcessInstanceContext.class);
+
+  private final Long processInstanceKey;
+  private final ValidationProvider validationProvider;
+  private final FeelEngineWrapper feelEngineWrapper;
+  private final ObjectMapper objectMapper;
+  private final Map<String, Object> properties;
+
+  public DefaultProcessInstanceContext(
+      final Long processInstanceKey,
+      final ValidationProvider validationProvider,
+      final FeelEngineWrapper feelEngineWrapper,
+      final ObjectMapper objectMapper,
+      final Map<String, Object> contextProperties,
+      final Map<String, Object> operateVariables) {
+    this.processInstanceKey = processInstanceKey;
+    this.validationProvider =
+        validationProvider == null
+            ? ValidationUtil.discoverDefaultValidationProviderImplementation()
+            : validationProvider;
+    this.feelEngineWrapper = feelEngineWrapper;
+    this.objectMapper = objectMapper;
+    this.properties = new HashMap<>(contextProperties);
+    contextProperties.putAll(operateVariables);
+  }
+
+  @Override
+  public <T> T bind(final Class<T> cls) {
+    try {
+      if (evaluateAndPrepareForBinding()) {
+        T mappedObject = objectMapper.convertValue(properties, cls);
+        validationProvider.validate(mappedObject);
+        return mappedObject;
+      } else {
+        LOGGER.debug(
+            "Evaluation and preparation for binding failed for processInstanceKey {}.",
+            processInstanceKey);
+        return null;
+      }
+    } catch (Exception e) {
+      LOGGER.warn(
+          "Error occurred while binding properties for processInstanceKey {}: {}",
+          processInstanceKey,
+          e.getMessage());
+      return null;
+    }
+  }
+
+  private boolean evaluateAndPrepareForBinding() {
+    for (Map.Entry<String, Object> entry : properties.entrySet()) {
+      String value = entry.getValue().toString();
+      if (isFeelExpression(value)) {
+        try {
+          Object evaluatedValue = feelEngineWrapper.evaluate(value, properties);
+          if (evaluatedValue == null) {
+            return false;
+          }
+          entry.setValue(evaluatedValue);
+        } catch (FeelEngineWrapperException e) {
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  private boolean isFeelExpression(String value) {
+    return value != null && value.trim().startsWith("=");
+  }
```

Comment:
We dont need this check.

Reply:
But we can bind values and evaluate properties already for inbound contexts. Can we re-use:
```
  public <T> T bindProperties(Class<T> cls) {
    var mappedObject = objectMapper.convertValue(getPropertiesWithSecrets(), cls);
    getValidationProvider().validate(mappedObject);
    return mappedObject;
  }
```

?

Reply:
Hi @Oleksiivanov I though we only bind against the inbound context variables and then (if needed) add the. process variables for the evaluation. For example:
```
@FEEL Function<Map<String,Object>, String> url;
```
and if you want to use the `url` you call url.apply(processContext.variables()). Maybe @chillleader or @igpetrov have other ideas but this kind of "pre-evaluation" mixes both concepts.

Reply:
Yes, I like the approach @sbuettner suggested, it would be nice to reuse the delayed evaluation concept. The only concern is that this might cause confusion if customers try to use `contextProperties` (i.e. defined directly in the element template of the polling connector) as variables in FEEL expression (as opposed to only using the process variables we fetched from Operate). On the other hand, I'm not sure if there is really a valid use case for this.

And, apparently, another issue with delayed evaluation: we can't reuse the HTTP request data models directly (the `HttpCommonRequest`) when binding properties, as we would have to wrap all fields into `@FEEL`-annotated `Function`.

Reply:
One more idea: what if we customized the effect of the `FEEL` annotation on the object mapper based on the connector type?
We could have 2 implementations of the `FeelDeserializer`, the first one would just be the current implementation that only uses process definition context to resolve variables, and the second would also include the process instance variables.

The first implementation could stay active by default, and it can be manually overridden in the `ProcessInstanceContext` by creating a local copy of the object mapper and enabling a modified jackson module on it. `ProcessInstanceContext` can pre-configure the custom jackson module by providing the variables from Operate.

Reply:
@chillleader We could then maybe start with a simpler approach:
* Remove the `bind` method on the `ProcessInstanceContext` for now
* Replace it with a `getVariables()` method
* Just use the existing infrastructure and bind against the inbound context plus add the variables when access needed?

WDYT?

Reply:
> Not necessarily, I think in theory we could use [deserialization context attributes](https://stackoverflow.com/a/46893967) to pass the process instance ID or even already resolved variables, thus allowing to have a common mapper instance for all ProcessInstanceContext objects.

That could work although it seems to create a new object reader instance for every process instance.

> So in the last suggestion, do I get it correctly that we would then add an overload of the bindProperties method to enrich the context with instance variables? Or how would you add them?
I would just add a `getVariables()` method to the `ProcessInstanceContext` and leave the existing `bindProperties` method as it is for now.

If we go this route we could discuss your idea in our architecture session tomorrow and plan it as an improvement. WDYT?